File: samples/NivaraInference/ResNet18.cs
using Nivara.AutoDiff;
using Nivara.AutoDiff.Nn;
using Nivara.AutoDiff.Operations;

namespace NivaraInference;

internal sealed class ResNet18 : Module<float>
{
    readonly Conv2d<float> stemConv;
    readonly BatchNorm2d<float> stemBn;
    readonly MaxPool2d<float> stemPool;
    readonly Module<float> stage0Layer0;
    readonly Module<float> stage0Layer1;
    readonly Module<float> stage1Layer0;
    readonly Module<float> stage1Layer1;
    readonly Module<float> stage2Layer0;
    readonly Module<float> stage2Layer1;
    readonly Module<float> stage3Layer0;
    readonly Module<float> stage3Layer1;
    readonly AdaptiveAvgPool2d<float> avgPool;
    readonly Linear<float> fc;

    public ResNet18(int numClasses = 1000)
    {
        stemConv = new Conv2d<float>(3, 64, 7, stride: 2, padding: 3, bias: false);
        stemBn = new BatchNorm2d<float>(64);
        stemPool = new MaxPool2d<float>(kernelSize: 3, stride: 2, padding: 1);

        stage0Layer0 = new BasicBlock(64, 64);
        stage0Layer1 = new BasicBlock(64, 64);

        stage1Layer0 = new BasicBlock(64, 128, stride: 2);
        stage1Layer1 = new BasicBlock(128, 128);

        stage2Layer0 = new BasicBlock(128, 256, stride: 2);
        stage2Layer1 = new BasicBlock(256, 256);

        stage3Layer0 = new BasicBlock(256, 512, stride: 2);
        stage3Layer1 = new BasicBlock(512, 512);

        avgPool = new AdaptiveAvgPool2d<float>(1);
        fc = new Linear<float>(512, numClasses, bias: true);

        RegisterModules(stemConv, stemBn, stemPool);
        RegisterModules(stage0Layer0, stage0Layer1);
        RegisterModules(stage1Layer0, stage1Layer1);
        RegisterModules(stage2Layer0, stage2Layer1);
        RegisterModules(stage3Layer0, stage3Layer1);
        RegisterModules(avgPool, fc);
    }

    public override ReverseGradTensor<float> Forward(ReverseGradTensor<float> input)
    {
        var x = stemConv.Forward(input);
        x = stemBn.Forward(x);
        x = ReverseGradOperations.Relu(x);
        x = stemPool.Forward(x);

        x = stage0Layer0.Forward(x);
        x = stage0Layer1.Forward(x);

        x = stage1Layer0.Forward(x);
        x = stage1Layer1.Forward(x);

        x = stage2Layer0.Forward(x);
        x = stage2Layer1.Forward(x);

        x = stage3Layer0.Forward(x);
        x = stage3Layer1.Forward(x);

        x = avgPool.Forward(x);

        int n = x.Shape[0], c = x.Shape[1];
        x.Reshape(n, c);

        x = fc.Forward(x);
        return x;
    }

    sealed class BasicBlock : Module<float>
    {
        internal readonly Conv2d<float> conv1;
        internal readonly BatchNorm2d<float> bn1;
        internal readonly Conv2d<float> conv2;
        internal readonly BatchNorm2d<float> bn2;
        internal readonly Conv2d<float>? downsampleConv;
        internal readonly BatchNorm2d<float>? downsampleBn;
        internal readonly bool hasDownsample;

        public BasicBlock(int inChannels, int outChannels, int stride = 1)
        {
            hasDownsample = inChannels != outChannels || stride != 1;

            conv1 = new Conv2d<float>(inChannels, outChannels, 3, stride: stride, padding: 1, bias: false);
            bn1 = new BatchNorm2d<float>(outChannels);

            conv2 = new Conv2d<float>(outChannels, outChannels, 3, padding: 1, bias: false);
            bn2 = new BatchNorm2d<float>(outChannels);

            if (hasDownsample)
            {
                downsampleConv = new Conv2d<float>(inChannels, outChannels, 1, stride: stride, bias: false);
                downsampleBn = new BatchNorm2d<float>(outChannels);
                RegisterModules(conv1, bn1, conv2, bn2, downsampleConv, downsampleBn);
            }
            else
            {
                RegisterModules(conv1, bn1, conv2, bn2);
            }
        }

        public override ReverseGradTensor<float> Forward(ReverseGradTensor<float> input)
        {
            var x = conv1.Forward(input);
            x = bn1.Forward(x);
            x = ReverseGradOperations.Relu(x);

            x = conv2.Forward(x);
            x = bn2.Forward(x);

            var residual = hasDownsample
                ? downsampleBn!.Forward(downsampleConv!.Forward(input))
                : input;

            x = x + residual;
            x = ReverseGradOperations.Relu(x);
            return x;
        }
    }

    internal static void LoadConv(Module<float> conv, float[] data, int[] shape)
    {
        var tensor = ReverseGradTensor<float>.FromMatrix(data, shape[0], shape[1] * shape[2] * shape[3]);
        conv.LoadStateDict(new Dictionary<string, ReverseGradTensor<float>> { ["Weight"] = tensor });
    }

    internal static void LoadBn(BatchNorm2d<float> bn,
        float[]? weight, float[]? bias, float[]? runningMean, float[]? runningVar)
    {
        var dict = new Dictionary<string, ReverseGradTensor<float>>();
        if (weight != null) dict["Weight"] = ReverseGradTensor<float>.FromArray(weight);
        if (bias != null) dict["Bias"] = ReverseGradTensor<float>.FromArray(bias);
        if (runningMean != null) dict["running_mean"] = ReverseGradTensor<float>.FromArray(runningMean);
        if (runningVar != null) dict["running_var"] = ReverseGradTensor<float>.FromArray(runningVar);
        if (dict.Count > 0) bn.LoadStateDict(dict);
    }

    internal static void LoadLinear(Linear<float> linear, float[] weightData, int[] weightShape, float[]? biasData)
    {
        var dict = new Dictionary<string, ReverseGradTensor<float>>
        {
            ["Weight"] = ReverseGradTensor<float>.FromMatrix(weightData, weightShape[0], weightShape[1])
        };
        if (biasData != null) dict["Bias"] = ReverseGradTensor<float>.FromMatrix(biasData, 1, biasData.Length);
        linear.LoadStateDict(dict);
    }

    public static int CountParameters(Dictionary<string, (float[] Data, int[] Shape)> tensors)
    {
        return tensors.Values.Sum(v => v.Data.Length);
    }

    static void LoadBasicBlock(Module<float> block, string prefix,
        Dictionary<string, (float[] Data, int[] Shape)> tensors)
    {
        var bb = (BasicBlock)block;

        LoadConv(bb.conv1, tensors[$"{prefix}.layer.0.convolution.weight"].Data,
            tensors[$"{prefix}.layer.0.convolution.weight"].Shape);
        LoadBn(bb.bn1,
            tensors.TryGetValue($"{prefix}.layer.0.normalization.weight", out var w1) ? w1.Data : null,
            tensors.TryGetValue($"{prefix}.layer.0.normalization.bias", out var b1) ? b1.Data : null,
            tensors.TryGetValue($"{prefix}.layer.0.normalization.running_mean", out var m1) ? m1.Data : null,
            tensors.TryGetValue($"{prefix}.layer.0.normalization.running_var", out var v1) ? v1.Data : null);

        LoadConv(bb.conv2, tensors[$"{prefix}.layer.1.convolution.weight"].Data,
            tensors[$"{prefix}.layer.1.convolution.weight"].Shape);
        LoadBn(bb.bn2,
            tensors.TryGetValue($"{prefix}.layer.1.normalization.weight", out var w2) ? w2.Data : null,
            tensors.TryGetValue($"{prefix}.layer.1.normalization.bias", out var b2) ? b2.Data : null,
            tensors.TryGetValue($"{prefix}.layer.1.normalization.running_mean", out var m2) ? m2.Data : null,
            tensors.TryGetValue($"{prefix}.layer.1.normalization.running_var", out var v2) ? v2.Data : null);

        if (bb.hasDownsample && bb.downsampleConv != null && bb.downsampleBn != null)
        {
            LoadConv(bb.downsampleConv, tensors[$"{prefix}.shortcut.convolution.weight"].Data,
                tensors[$"{prefix}.shortcut.convolution.weight"].Shape);
            LoadBn(bb.downsampleBn,
                tensors.TryGetValue($"{prefix}.shortcut.normalization.weight", out var sw) ? sw.Data : null,
                tensors.TryGetValue($"{prefix}.shortcut.normalization.bias", out var sb) ? sb.Data : null,
                tensors.TryGetValue($"{prefix}.shortcut.normalization.running_mean", out var sm) ? sm.Data : null,
                tensors.TryGetValue($"{prefix}.shortcut.normalization.running_var", out var sv) ? sv.Data : null);
        }
    }

    public static ResNet18 LoadWeights(Dictionary<string, (float[] Data, int[] Shape)> tensors, int numClasses = 1000)
    {
        var model = new ResNet18(numClasses);
        model.Eval();

        LoadConv(model.stemConv,
            tensors["resnet.embedder.embedder.convolution.weight"].Data,
            tensors["resnet.embedder.embedder.convolution.weight"].Shape);
        LoadBn(model.stemBn,
            tensors.TryGetValue("resnet.embedder.embedder.normalization.weight", out var sw0) ? sw0.Data : null,
            tensors.TryGetValue("resnet.embedder.embedder.normalization.bias", out var sb0) ? sb0.Data : null,
            tensors.TryGetValue("resnet.embedder.embedder.normalization.running_mean", out var sm0) ? sm0.Data : null,
            tensors.TryGetValue("resnet.embedder.embedder.normalization.running_var", out var sv0) ? sv0.Data : null);

        LoadBasicBlock(model.stage0Layer0, "resnet.encoder.stages.0.layers.0", tensors);
        LoadBasicBlock(model.stage0Layer1, "resnet.encoder.stages.0.layers.1", tensors);
        LoadBasicBlock(model.stage1Layer0, "resnet.encoder.stages.1.layers.0", tensors);
        LoadBasicBlock(model.stage1Layer1, "resnet.encoder.stages.1.layers.1", tensors);
        LoadBasicBlock(model.stage2Layer0, "resnet.encoder.stages.2.layers.0", tensors);
        LoadBasicBlock(model.stage2Layer1, "resnet.encoder.stages.2.layers.1", tensors);
        LoadBasicBlock(model.stage3Layer0, "resnet.encoder.stages.3.layers.0", tensors);
        LoadBasicBlock(model.stage3Layer1, "resnet.encoder.stages.3.layers.1", tensors);

        LoadLinear(model.fc,
            tensors["classifier.1.weight"].Data,
            tensors["classifier.1.weight"].Shape,
            tensors.TryGetValue("classifier.1.bias", out var bias) ? bias.Data : null);

        return model;
    }
}

File: samples/NivaraInference/Python/resnet18.py
"""ResNet-18 inference using HuggingFace SafeTensors weights (same as C# sample)."""
import time
import sys
import os
import torch
import torch.nn as nn

sys.path.insert(0, os.path.dirname(__file__))
from hf_loader import load_safetensors, IMAGES_DIR, MODELS_DIR, print_top_k

MEAN = [0.485, 0.456, 0.406]
STD = [0.229, 0.224, 0.225]


class BasicBlock(nn.Module):
    def __init__(self, in_ch, out_ch, stride=1):
        super().__init__()
        self.conv1 = nn.Conv2d(in_ch, out_ch, 3, stride=stride, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(out_ch)
        self.conv2 = nn.Conv2d(out_ch, out_ch, 3, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(out_ch)
        self.relu = nn.ReLU(inplace=True)
        self.has_downsample = in_ch != out_ch or stride != 1
        if self.has_downsample:
            self.downsample = nn.Sequential(
                nn.Conv2d(in_ch, out_ch, 1, stride=stride, bias=False),
                nn.BatchNorm2d(out_ch),
            )
        else:
            self.downsample = None

    def forward(self, x):
        identity = x
        out = self.relu(self.bn1(self.conv1(x)))
        out = self.bn2(self.conv2(out))
        if self.downsample is not None:
            identity = self.downsample(x)
        return self.relu(out + identity)


class ResNet18HF(nn.Module):
    def __init__(self, num_classes=1000):
        super().__init__()
        self.stem_conv = nn.Conv2d(3, 64, 7, stride=2, padding=3, bias=False)
        self.stem_bn = nn.BatchNorm2d(64)
        self.stem_pool = nn.MaxPool2d(3, stride=2, padding=1)
        self.relu = nn.ReLU(inplace=True)

        self.stage0_layer0 = BasicBlock(64, 64)
        self.stage0_layer1 = BasicBlock(64, 64)
        self.stage1_layer0 = BasicBlock(64, 128, stride=2)
        self.stage1_layer1 = BasicBlock(128, 128)
        self.stage2_layer0 = BasicBlock(128, 256, stride=2)
        self.stage2_layer1 = BasicBlock(256, 256)
        self.stage3_layer0 = BasicBlock(256, 512, stride=2)
        self.stage3_layer1 = BasicBlock(512, 512)

        self.avgpool = nn.AdaptiveAvgPool2d(1)
        self.fc = nn.Linear(512, num_classes)

    def forward(self, x):
        x = self.relu(self.stem_bn(self.stem_conv(x)))
        x = self.stem_pool(x)
        x = self.stage0_layer0(x)
        x = self.stage0_layer1(x)
        x = self.stage1_layer0(x)
        x = self.stage1_layer1(x)
        x = self.stage2_layer0(x)
        x = self.stage2_layer1(x)
        x = self.stage3_layer0(x)
        x = self.stage3_layer1(x)
        x = self.avgpool(x)
        x = torch.flatten(x, 1)
        x = self.fc(x)
        return x


def load_weights(model, tensors):
    """Load weights from HuggingFace safetensors into the PyTorch model."""
    sd = model.state_dict()

    # Mapping: HF key -> PyTorch key
    mapping = {
        "resnet.embedder.embedder.convolution.weight": "stem_conv.weight",
        "resnet.embedder.embedder.normalization.weight": "stem_bn.weight",
        "resnet.embedder.embedder.normalization.bias": "stem_bn.bias",
        "resnet.embedder.embedder.normalization.running_mean": "stem_bn.running_mean",
        "resnet.embedder.embedder.normalization.running_var": "stem_bn.running_var",
        "resnet.embedder.embedder.normalization.num_batches_tracked": "stem_bn.num_batches_tracked",
    }

    for stage in range(4):
        for layer in range(2):
            for k in range(2):
                prefix = f"resnet.encoder.stages.{stage}.layers.{layer}"
                py_layer = f"stage{stage}_layer{layer}"
                mapping[f"{prefix}.layer.{k}.convolution.weight"] = f"{py_layer}.conv{k+1}.weight"
                mapping[f"{prefix}.layer.{k}.normalization.weight"] = f"{py_layer}.bn{k+1}.weight"
                mapping[f"{prefix}.layer.{k}.normalization.bias"] = f"{py_layer}.bn{k+1}.bias"
                mapping[f"{prefix}.layer.{k}.normalization.running_mean"] = f"{py_layer}.bn{k+1}.running_mean"
                mapping[f"{prefix}.layer.{k}.normalization.running_var"] = f"{py_layer}.bn{k+1}.running_var"
                mapping[f"{prefix}.layer.{k}.normalization.num_batches_tracked"] = f"{py_layer}.bn{k+1}.num_batches_tracked"

            # shortcut
            prefix = f"resnet.encoder.stages.{stage}.layers.{layer}"
            py_layer = f"stage{stage}_layer{layer}"
            mapping[f"{prefix}.shortcut.convolution.weight"] = f"{py_layer}.downsample.0.weight"
            mapping[f"{prefix}.shortcut.normalization.weight"] = f"{py_layer}.downsample.1.weight"
            mapping[f"{prefix}.shortcut.normalization.bias"] = f"{py_layer}.downsample.1.bias"
            mapping[f"{prefix}.shortcut.normalization.running_mean"] = f"{py_layer}.downsample.1.running_mean"
            mapping[f"{prefix}.shortcut.normalization.running_var"] = f"{py_layer}.downsample.1.running_var"
            mapping[f"{prefix}.shortcut.normalization.num_batches_tracked"] = f"{py_layer}.downsample.1.num_batches_tracked"

    mapping["classifier.1.weight"] = "fc.weight"
    mapping["classifier.1.bias"] = "fc.bias"

    loaded = 0
    for hf_key, pt_key in mapping.items():
        if hf_key in tensors and pt_key in sd:
            sd[pt_key] = tensors[hf_key]
            loaded += 1

    model.load_state_dict(sd)
    print(f"  Loaded {loaded}/{len(mapping)} parameters")


def preprocess_image(path, size=224):
    """Load image, resize to size×size, apply ImageNet normalization."""
    from PIL import Image
    import torchvision.transforms as T
    transform = T.Compose([
        T.Resize(size),
        T.ToTensor(),
        T.Normalize(mean=MEAN, std=STD),
    ])
    img = Image.open(path).convert("RGB")
    return transform(img).unsqueeze(0)


def main():
    print("=== HuggingFace ResNet-18 Inference (same weights as C#) ===")
    print()

    model_dir = os.path.join(MODELS_DIR, "resnet18")
    print(f"Loading weights from {model_dir}/model.safetensors...")
    tensors = load_safetensors(model_dir)
    print(f"  Loaded {len(tensors)} tensors")

    print("Building ResNet-18 model...")
    model = ResNet18HF(num_classes=1000)
    model.eval()
    load_weights(model, tensors)

    param_count = sum(p.numel() for p in model.parameters())
    print(f"  Parameters: {param_count:,}")
    print()

    # Warmup
    print("Warmup (3 passes)...")
    dummy = torch.randn(1, 3, 224, 224)
    with torch.no_grad():
        for _ in range(3):
            model(dummy)
    print()

    # Benchmark synthetic
    print("Benchmark: synthetic 224x224 input (10 passes)...")
    times = []
    for i in range(10):
        inp = torch.randn(1, 3, 224, 224)
        start = time.perf_counter()
        with torch.no_grad():
            out = model(inp)
        elapsed = (time.perf_counter() - start) * 1000
        times.append(elapsed)
        print(f"  Run {i+1:2d}: {elapsed:.1f} ms")

    avg = sum(times) / len(times)
    print(f"  Average: {avg:.1f} ms  (min={min(times):.1f}, max={max(times):.1f})")
    print()
    print_top_k(out)
    print()

    # Real images
    if os.path.exists(IMAGES_DIR):
        image_files = sorted([f for f in os.listdir(IMAGES_DIR) if f.endswith(".jpg")])
        if image_files:
            print("Benchmark: real images...")
            for fname in image_files:
                path = os.path.join(IMAGES_DIR, fname)
                from PIL import Image
                img = Image.open(path)
                inp = preprocess_image(path)
                start = time.perf_counter()
                with torch.no_grad():
                    out = model(inp)
                elapsed = (time.perf_counter() - start) * 1000
                print(f"  {fname} ({img.size[0]}x{img.size[1]}): {elapsed:.1f} ms")
                print_top_k(out, k=3)
                print()


if __name__ == "__main__":
    main()
